### PR TITLE
feat: Add support to return the ESXi host and vCenter Server password policy

### PR DESCRIPTION
**Pull request by garlicNova**
In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/blob/main/CONTRIBUTING.md) for making a pull request.

**Summary of Pull Request**

Add 2 cmdlets
- `Get-ESXiPasswordPolicy`
- `Get-VCServerPasswordPolicy`

This cmdlets are used return the password expiration policy for an ESXi host and vCenter Server instance in number of days.

Signed-off-by: Kevin Teng <kteng@vmware.com>

**Type of Pull Request**

- [ ] This is a bug fix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.
      Please describe:

**Related to Existing Issues**

Issue Number: #31 

**Test and Documentation Coverage**

- [x] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

**Review by GaryJBlake — DISMISSED** [review]

File: PowerValidatedSolutions.psm1 [hunk 1]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$cluster
+	)
+	
+	# construct ESXi server Object
+	$esxiPasswdPolicy = New-Object System.Collections.Generic.List[System.Object]
+	
+	Try{
+		if (Test-Connection -server $server) {
+			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+				if (Get-VCFWorkloadDomain | Where-Object {$_.name -eq $domain}) {
```
[the remark this answers is not in the document]
Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:459]: Good catch, Gary.

File: PowerValidatedSolutions.psm1 [hunk 2]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+											}
+										}
+										return $esxiPasswdPolicy
+									}
+									else {
+										Write-Warning "No ESXi hosts found within $cluster cluster."
+									}
+								}
+								else {
+										Write-Error "Cluster $cluster not found in $vcfVcenterDetails.fqdn vCenter Server."
```
[the remark this answers is not in the document]
Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:491]: Updated the markdown in the review.

File: PowerValidatedSolutions.psm1 [hunk 3]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	Catch{
```
[the remark this answers is not in the document]
Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:500]: Updated the markdown in the review.

File: PowerValidatedSolutions.psm1 [hunk 4]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$server,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$user,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$pass,
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$domain
+	)
+
+	Try{
+		if (Test-Connection -server $server) {
+			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+				if (Get-VCFWorkloadDomain | Where-Object {$_.name -eq $domain}) {
```
[the remark this answers is not in the document]
Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:534]: Good catch.

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Great review by Gary - _thank you!_

@garlicNova, can you make these minor changes and also ensure you update your git config with your `user.name` and `user.email` before the commit?

```bash
git config --global user.email kteng@vmware.com
git config --global user.name Kevin Teng
```

File: PowerValidatedSolutions.psm1 [hunk 5]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
 }
 Export-ModuleMember -Function Undo-SddcManagerRole
 
+Function Get-EsxiPasswordPolicy {
+	<#
+        .SYNOPSIS
+        Retrieve ESXi servers' Password Policies
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:428]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 6]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+
+        .DESCRIPTION
+        The Get-EsxiPasswordPolicy cmdlet returns a list contains a list of ESXi
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:431]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 7]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+		The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that the workload domain exists in the SDDC Manager inventory
+        - Validates that network connectivity and authentication is possible to vCenter Server
+        - Gathers the ESXi hosts for the cluster specificed
+        - Retrieve all ESXi hosts password policies
+
+        .EXAMPLE
+        Get-EsxiPasswordPolicy -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -cluster sfo-m01-cl0l
+        This example retrieves all ESXi hosts password policies within the cluster named sfo-m01-cl01 of the workload domain sfo-m01
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:442]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 8]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+
+    #>
+	Param(
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:445]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 9]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$server,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$user,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$pass,
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$domain,
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$cluster
+	)
+	
+	# construct ESXi server Object
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:453]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 10]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+}
+Export-ModuleMember -Function Get-EsxiPasswordPolicy
+
+Function Get-VCServerPasswordPolicy {
+	<#
+        .SYNOPSIS
+        Retrieve vCenter servers' Password Policy
+
+        .DESCRIPTION
+        The Get-VCServerPasswordPolicy cmdlet returns the current configured password expiration period
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:512]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 11]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$user,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$pass,
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$domain,
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$cluster
+	)
+	
+	# construct ESXi server Object
+	$esxiPasswdPolicy = New-Object System.Collections.Generic.List[System.Object]
+	
+	Try{
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:456]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:491]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:459]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:500]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 12]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+	Catch{
+		Debug-ExceptionWriter -object $_
+	}
+}
+Export-ModuleMember -Function Get-EsxiPasswordPolicy
+
+Function Get-VCServerPasswordPolicy {
+	<#
+        .SYNOPSIS
+        Retrieve vCenter servers' Password Policy
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:509]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 13]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+
+        .DESCRIPTION
+        The Get-VCServerPasswordPolicy cmdlet returns the current configured password expiration period
+		and the email for warning notification settings for the vCenter Server root account
+        The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to vCenter Server
+		- Retrieve the password expiration period and the email for warning notification settings for the vCenter Server root account
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:517]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 14]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+
+        .EXAMPLE
+        Get-VCServerPasswordPolicy -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01
+        This example retrieves the password expiration period and the email for warning notification settings for the vCenter Server root account
+
+    #>
+	Param(
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:524]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 [hunk 15]
```diff
@@ -422,6 +422,152 @@ Function Undo-SddcManagerRole {
...
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$server,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$user,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$pass,
+		[Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$domain
+	)
+
+	Try{
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:531]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

File: PowerValidatedSolutions.psm1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (garlicNova) [reply PowerValidatedSolutions.psm1:534]: addressed in fffa0d4c3b50c0571ea7756dea0f448fc0ad6508

**Review by tenthirtyam — APPROVED** [review]
LGTM! 🚀

**Pull request merged**